sam/app-sfn-kms-integration/template.yaml

```diff
@@ -0,0 +1,187 @@
+        DDBPutItem: !Sub arn:${AWS::Partition}:states:::dynamodb:putItem
+        DDBTable: !Ref TransactionTable
+      Definition:
+        Comment: A state machine that does mock stock trading.
+        StartAt: Check Stock Value
+        States:
+          Check Stock Value:
+            Type: Task
+            Resource: "${StockCheckerFunctionArn}"
+            Next: Buy or Sell?
+          Buy or Sell?:
+            Type: Choice
+            Choices:
+            - Variable: "$.stock_price"
+              NumericLessThanEquals: 50
+              Next: Buy Stock
+            Default: Sell Stock
+          Sell Stock:
+            Type: Task
+            Resource: "${StockSellerFunctionArn}"
+            Next: Record Transaction
+          Buy Stock:
+            Type: Task
+            Resource: "${StockBuyerFunctionArn}"
+            Next: Record Transaction
+          Record Transaction:
+            Type: Task
+            Resource: "${DDBPutItem}"
+            Parameters:
+              TableName: "${DDBTable}"
+              Item:
+                Id:
+                  S.$: "$.id"
+                Type:
+                  S.$: "$.type"
+                Price:
+                  N.$: "$.price"
+                Quantity:
+                  N.$: "$.qty"
+                Timestamp:
+                  S.$: "$.timestamp"
+            Retry:
+            - ErrorEquals:
+              - States.TaskFailed
+              IntervalSeconds: 20
+              MaxAttempts: 5
+              BackoffRate: 10
+            End: true
+
+  StocksKmsKey:
+    Type: AWS::KMS::Key
+    Properties:
+      Description: Symmetric KMS key used for encryption/decryption provided by customer
+      KeyPolicy:
+        Version: "2012-10-17"
+        Statement:
+          - Effect: Allow
+            Principal:
+              AWS: !Sub arn:aws:iam::${AWS::AccountId}:root
+            Action:
+              - kms:Create*
+              - kms:Describe*
+              - kms:Enable*
+              - kms:List*
+              - kms:Put*
+              - kms:Update*
+              - kms:Revoke*
+              - kms:Disable*
+              - kms:Get*
+              - kms:Delete*
+              - kms:ScheduleKeyDeletion
+              - kms:CancelKeyDeletion
+            Resource: '*'
+          - Effect: Allow
+            Principal:
+              AWS: !GetAtt StockTradingStateMachineExecutionRole.Arn
+            Action:
+              - kms:GenerateDataKey*
+              - kms:Decrypt
+            Resource: '*'
+            Condition:
+              StringEquals:
+                kms:EncryptionContext:aws:states:stateMachineArn: !Join 
+                  - ":"
+                  - - !Sub arn:${AWS::Partition}:states
+                    - !Ref AWS::Region
+                    - !Ref AWS::AccountId
+                    - stateMachine
+                    - !FindInMap ['StateMachine', 'Name', 'Value']
+
+  StockTradingStateMachineExecutionRole:
```

Comment:
Also make sure that the execution role has the following permission when using KMS encryption. Otherwise you won't be able to start/stop executions using the execution role.

```
StartExecution (API)
kms:GenerateDataKey
kms:Decrypt
```
```
StopExecution (API)
kms:Decrypt
```

Reply:
Shouldn't the key policy be sufficient enough because the sfn execution role is categorically granted access to these two actions? @VaidSaraswat